Mathlib/Algebra/Order/Quantale.lean

```diff
@@ -0,0 +1,161 @@
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+variable {α : Type*} {ι : Type*} {x y z : α} {s : Set α} {f : ι → α}
+variable [Semigroup α] [CompleteLattice α] [IsQuantale α]
+
+@[to_additive]
+theorem mul_sSup_eq_iSup_mul : x * sSup s = ⨆ y ∈ s, x * y := IsQuantale.mul_sSup_eq_iSup_mul _ _
+
+@[to_additive]
+theorem sSup_mul_eq_iSup_mul : sSup s * x = ⨆ y ∈ s, y * x := IsQuantale.sSup_mul_eq_iSup_mul _ _
+
+@[to_additive]
+theorem mul_iSup_distrib : x * ⨆ i, f i = ⨆ i, x * f i := by
+  rw [iSup, mul_sSup_eq_iSup_mul, iSup_range]
+
+@[to_additive]
+theorem iSup_mul_distrib : (⨆ i, f i) * x = ⨆ i, f i * x := by
+  rw [iSup, sSup_mul_eq_iSup_mul, iSup_range]
+
+@[to_additive]
+theorem mul_sup_distrib : x * (y ⊔ z) = (x * y) ⊔ (x * z) := by
+  rw [← iSup_pair, ← sSup_pair, mul_sSup_eq_iSup_mul]
+
+@[to_additive]
+theorem sup_mul_distrib : (x ⊔ y) * z = (x * z) ⊔ (y * z) := by
+  rw [← (@iSup_pair _ _ _ (fun _? => _? * z) _ _), ← sSup_pair, sSup_mul_eq_iSup_mul]
+
+@[to_additive]
+instance : MulLeftMono α where
+  elim := by
+    intro _ _ _; simp only; intro
+    rw [← left_eq_sup, ← mul_sup_distrib, sup_of_le_left]
+    trivial
+
+@[to_additive]
+instance : MulRightMono α where
+  elim := by
+    intro _ _ _; simp only; intro
+    rw [← left_eq_sup, ← sup_mul_distrib, sup_of_le_left]
+    trivial
+
+/-- Left- and right-residuation operators on an additive quantale are similar to the Heyting
+operator on complete lattices, but for a non-commutative logic.
+I.e. `x ≤ y ⇨ₗ z ↔ x * y ≤ z` or alternatively `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+@[to_additive "Left- and right- residuation operators on an additive quantale are similar to
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ≤ y ⇨ₗ z ↔ x + y ≤ z` or alternatively `x ⇨ₗ y = sSup { z | z + x ≤ y }`.
+"]
+def leftMulResiduation (x y : α) := sSup {z | z * x ≤ y}
+
+/-- Left- and right- residuation operators on an additive quantale are similar to the Heyting
+operator on complete lattices, but for a non-commutative logic.
+I.e. `x ≤ y ⇨ᵣ z ↔ y * x ≤ z` or alternatively `x ⇨ₗ y = sSup { z | x * z ≤ y }`.
+-/
+@[to_additive "Left- and right- residuation operators on an additive quantale are similar to
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ≤ y ⇨ᵣ z ↔ y + x ≤ z` or alternatively `x ⇨ₗ y = sSup { z | x + z ≤ y }`.
+"]
+def rightMulResiduation (x y : α) := sSup {z | x * z ≤ y}
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ₗ " => leftMulResiduation
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ᵣ " => rightMulResiduation
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ₗ " => leftAddResiduation
+
+@[inherit_doc]
+scoped infixr:60 " ⇨ᵣ " => rightAddResiduation
```

Comment:
@YaelDillies how does Lean know which one (additive or multiplicative) to pick here?

Reply:
I think because I named them Mul and Add respectively.
For clarity, I now updated this and made the namespace explicit.

Reply:
Oh, I thought they were in a different namespace, in which case both notation should usually not be available at the same time

Reply:
It appears that having them in the same namespace was the mistake that actually made it work. I think Lean knows which ones to pick based on type inferrence. When I got them into different namespaces, it started having trouble because to_additive was not able to deduce the namespace.